kube-controllers/pkg/controllers/node/ipam.go

```diff
@@ -1074,6 +1170,140 @@ func (c *ipamController) kubernetesNodeForCalico(cnode string) (string, error) {
+		Name: "ipam_allocations_borrowed",
+		Help: "IPs currently allocated in IPAM to a workload or tunnel endpoint, where the allocation was borrowed " +
+			"from a block affine to another node.",
+		ConstLabels: prometheus.Labels{"ippool": poolName},
+	}, []string{"node"})
+	prometheus.MustRegister(borrowedAllocationGauges[poolName])
+
+	blocksGauges[poolName] = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Name:        "ipam_blocks",
+		Help:        "IPAM blocks currently allocated for the IP pool.",
+		ConstLabels: prometheus.Labels{"ippool": poolName},
+	}, []string{"node"})
+	prometheus.MustRegister(blocksGauges[poolName])
+
+	gcCandidateGauges[poolName] = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "ipam_allocations_gc_candidates",
+		Help: "Allocations that are currently marked by the garbage collector as potential candidates to " +
+			"reclaim. Under normal operation, this metric should return to zero after the garbage collector " +
+			"confirms that this allocation can be reclaimed and reclaims it, or the allocation is confirmed as valid.",
+		ConstLabels: prometheus.Labels{"ippool": poolName},
+	}, []string{"node"})
+	prometheus.MustRegister(gcCandidateGauges[poolName])
+
+	gcReclamationCounters[poolName] = prometheus.NewCounterVec(prometheus.CounterOpts{
+		Name: "ipam_allocations_gc_reclamations",
+		Help: "The total allocations that have been reclaimed by the garbage collector over time. Under normal " +
+			"operation, this counter should increase, and increases of this counter should align to a return to zero " +
+			"for the candidate gauge.",
+		ConstLabels: prometheus.Labels{"ippool": poolName},
+	}, []string{"node"})
+	prometheus.MustRegister(gcReclamationCounters[poolName])
+}
+
+func unregisterMetricVectorsForPool(poolName string) {
+	if _, ok := inUseAllocationGauges[poolName]; ok {
+		prometheus.Unregister(inUseAllocationGauges[poolName])
+		delete(inUseAllocationGauges, poolName)
+	}
+
+	if _, ok := borrowedAllocationGauges[poolName]; ok {
+		prometheus.Unregister(borrowedAllocationGauges[poolName])
+		delete(borrowedAllocationGauges, poolName)
+	}
+
+	if _, ok := blocksGauges[poolName]; ok {
+		prometheus.Unregister(blocksGauges[poolName])
+		delete(blocksGauges, poolName)
+	}
+
+	if _, ok := gcCandidateGauges[poolName]; ok {
+		prometheus.Unregister(gcCandidateGauges[poolName])
+		delete(gcCandidateGauges, poolName)
+	}
+
+	if _, ok := gcReclamationCounters[poolName]; ok {
+		prometheus.Unregister(gcReclamationCounters[poolName])
+		delete(gcReclamationCounters, poolName)
+	}
+}
+
+// Creates map used to index gauge values by node, and seeds with zeroes to create explicit zero values rather than
+// absence of data for a node. This enables users to construct utilization expressions that return 0 when the numerator
+// is zero, rather than no data. If the pool is the 'unknown' pool, the map is not seeded.
+func (c *ipamController) createZeroedMapForNodeValues(poolName string) map[string]int {
```

Comment:
Hm, I'm a little bit wary of this method of initializing the maps. 

I don't know that we can be 100% confident that the `kubernetesNodesByCalicoName` map includes all of the host names that might show up in an allocation's `Node` field. 

For example, allocations might have exist on nodes that don't exist in the k8s API, or we might have received an allocation update prior to learning about the node.

So, I think ultimately we still need a check in the `updateMetrics` function like this:

```go
if _, ok := xByNode[a.node()]; !ok {
  xByNode[a.node()] = 0
}
xByNode[a.node()]++
```

You could stick that within a helper though. For example, rather than a helper function to initialize a map, you could have a helper function to increment the counter.

```go
func incrementPerNodeMap(m map[string]int, node)
```

Reply:
I guess you might want this anyway to ensure that nodes with no allocations show up with `0` in the prometheus metrics, so perhaps you need both (an initialization helper, as well as some logic in `updateMetrics` to add nodes that we might not have initialized in this loop.